### PR TITLE
fw_pos_control_l1: apply pitch setpoint offsets centrally

### PR DESCRIPTION
**Pull request by dagar**
From https://github.com/PX4/Firmware/issues/11785.

> Negative pitch setpoint - If you have a negative pitch setpoint to use during level flight, you are essentially wanting the plane to nose down (or up) to not change altitude while in the air. The issue is that during takeoff if you look at pitch setpoint you will see that it is actually the specified pitch minus the setpoint (we have takeoff pitch set to 12º but are getting 7º of pitch up on takeoff because we have a negative set point of 5 degrees (FW_PSP_OFF). What we are seeing is a real “airframe” pitch of 7 degrees instead of 12 (because of the offset and it thinks it is 12º) and the elevator up is sort of giving 12º up but not really since when level on the ground the elevator is technically giving 5º pitch down.
> Quick solution is probably to simply let runway takeoff pitch setpoint (RWTO_PSP) able to be a positive number. It is currently not allowed to be.

The pitch and pitch limits are shifted by the pitch setpoint offset before sent into TECS. This was being done inconsistently only for position and loiter setpoint types.

Related https://github.com/PX4/Firmware/pull/7098

### CODE REVIEW

**Comment by dagar** [comment]
@ryanjAA

**Comment by dagar** [comment]
Since this PR will require testing I'm going to pull https://github.com/PX4/Firmware/pull/11373 in as well.

**Comment by dagar** [comment]
> This change breaks for users who were relying on the takeoff pitch field in QGC to set up their mission. That it is now being ignored comes unexpected. This would need to be coordinated with a QGC release. I would instead let the parameter override the mission parameter if it is non-zero.

If ignoring the existing mavlink parameter is unacceptable then we'd need more than a coordinated release to prevent potential confusion. How about we reject the mission upload if the field isn't 0 (or NAN) with a warning?

**Comment by LorenzMeier** [comment]
That’s fine with me.

**Comment by LorenzMeier** [comment]
@dagar SITL build failure

**Comment by dagar** [comment]
Needs to be limited to FW only (and has a typo). 

    [ERROR] [1555181955.941107127, 30.000000000]: FCU: Takeoff rejected, remove deprecated minium pitch
    [ERROR] [1555181955.994412932, 30.056000000]: FCU: IGN MISSION_ITEM: Busy
    [ERROR] [1555181956.939554436, 31.004000000]: WP: upload failed: param1 has an invalid value

**Comment by ryanjAA** [comment]
To understand the change in basic terms. If we have the following assumption:

Pitch offset of negative 11 degrees (fw_psp_off)
Runway way takeoff pitch min set via parameter of 10 degrees positive (RWTO_MAX_PITCH)

Qgc takeoff pitch is set to 12 degrees. 

Then expected pitch should be 21 degrees from the autopilots perspective (+11 for compensating from the negative psp offset and +10 for takeoff). and the qgc takeoff pitch is ignored. I’m assuming minimum pitch takeoff is is ignored although that doesn’t really matter since it would likely be so much higher. What about the max pitch? Is that going to need to be increased?
  

Am I understanding this correctly?

**Comment by LorenzMeier** [comment]
@ryanjAA Yes, although you would not be told by QGC that you need to set takeoff pitch to 0 and adjust it via the new param.

**Comment by ryanjAA** [comment]
Ok thanks, so it only will work if the takeoff pitch is 0 (in qgc mission planning) when creating and uploading the mission (from @dagar note above). I assume the mission simply wont upload (from above). I'll give this a try in SITL.

**Comment by dagar** [comment]
Actually, this is an important point to discuss. In current master there's a rather confusing mix of some of these pitch angle parameters being applied relative to FW_PSP_OFF, and some treated as absolute.

In the fixedwing position controller we have the following user configurable pitch settings in degrees.

 * FW_P_LIM_MIN
 * FW_P_LIM_MAX
 * FW_TKO_PITCH_MIN
 * FW_LND_ANG
 * FW_LND_FL_PMIN
 * FW_LND_FL_PMAX
 * FW_MAN_P_MAX

 Plus
 * FW_PSP_OFF

I would propose that all of these should be interpreted absolute, and not relative to FW_PSP_OFF. The FW_PSP_OFF parameter is then simply the expected pitch trim at cruise. 

@ryanjAA @Antiheavy your input here would be greatly appreciated.

**Comment by Antiheavy** [comment]
> I would propose that all of these should be interpreted absolute, and not relative to FW_PSP_OFF. The FW_PSP_OFF parameter is then simply the expected pitch trim at cruise.

This is basically what the parameter description says, so I guess that makes sense. https://github.com/PX4/Firmware/blob/20f870137b66c61df0ba0dcd6a96a8deab01f92e/src/modules/fw_att_control/fw_att_control_params.c#L433-L446

Can someone describe how FW_PSP_OFF is actually used in the controller?

**Comment by Antiheavy** [comment]
A related question is:  does FW_PSP_OFF add on top of the absolute limits, such as "FW_P_LIM_MAX"?  I think the answer is it should be allowed to add to the absolute limits, since it is an offset.

The use case I image is you have a fixed wing design where you'd like the lifting surfaces pitch datum to be different than the avionics system pitch datum.

**Comment by dagar** [comment]
> A related question is: does FW_PSP_OFF add on top of the absolute limits, such as "FW_P_LIM_MAX"? I think the answer is it should be allowed to add to the absolute limits, since it is an offset.

Those limits are absolute. That's the change that created the different interpretations. https://github.com/PX4/Firmware/pull/7098

At this point I think I could go either way as long as it's consistent and documented.

**Comment by Antiheavy** [comment]
> Those limits are absolute. That's the change that created the different interpretations. #7098
> 
> At this point I think I could go either way as long as it's consistent and documented.

okay, so is the following correct (as of current state of this PR)?
    FW_P_LIM_MIN --> absolute to avionics coordinate system, cannot be breached by FW_PSP_OFF.
    FW_P_LIM_MAX  --> absolute to avionics coordinate system, cannot be breached by FW_PSP_OFF.
    FW_TKO_PITCH_MIN  --> modified by FW_PSP_OFF.
    FW_LND_ANG -->  hmm.... does this apply?
    FW_LND_FL_PMIN   --> modified by FW_PSP_OFF.
    FW_LND_FL_PMAX   --> modified by FW_PSP_OFF.
    FW_MAN_P_MAX  --> absolute to avionics coordinate system, cannot be breached by FW_PSP_OFF.

**Comment by Antiheavy** [comment]
Should we consider updating the short description for FW_PSP_OFF to "Pitch setpoint offset at cruise"?

**Comment by Antiheavy** [comment]
The Takeoff minimum pitch stuff looks correct to me.

**Comment by ryanjAA** [comment]
I’m really glad to see this getting cleared up. It never made sense when we started talking about it why a cruise speed offset was applying that offset to everything. From conversations with @dagar tecs initializing from the offset and not zero just didn’t make sense but made sense why so many things weren’t acting as expected. Countless logs with shallow takeoffs and then finally when the exact difference was the offset it alll started to piece together. Definitely on the landing side as well. Now (if I’m understanding correctly), except for at cruise, there is no offset applied. All are absolute which is so much better. Landing with a minimum flare of 3 and offset of -5 meant a -2 degree min flare which never made sense not made it in the docs which would explain a lot and why min flare had to often be exaggerated. A description as written by @dagar and noted by @antiheavy is a must and I think if there are people out there relying on an offset they need to see the difference (not sure how that part will happen) since functionality is changed. On a side note and after today’s talk with @dagar, why the autopilot isn’t figuring out what the cruise trim offset needs to be automatically is another question but one thing at a time I suppose. 

Glad to see this getting changed.

**Comment by LorenzMeier** [comment]
@ryanjAA Could you test this today?

**Comment by ryanjAA** [comment]
I can’t say definitively since we have another storm and 50 km/h winds headed our way (weather is messing with our flight schedule lately). We may take one out for good measure but going depend on thunder and lightning which as of now looks likely. If by miracle we get in the air I’ll have logs but can’t confirm as of now. Sending someone out there though in 2 hours so will be on weather standby.

**Comment by LorenzMeier** [comment]
@ryanjAA We're getting really close to the release cut-off date. How does testing look like here?

**Comment by ryanjAA** [comment]
Doesn’t look good right now. We had wind and have thunderstorms approaching and Easter is annoyingly in the way. 

I sent someone out on Thursday but they registered 20.5 knot winds on the ground with the anemometer. They still took off for a baseline given the wind prior to the pr firmware since technically the foam ones can handle that (its at the very edge if heavy but it’s done it quite a lot) but had 28 knot sustained once back on the ground which only the composites can do without any worry (and there wasn’t one there for testing since the PR was pitch based and our safety protocol makes it so we have to have 2 minimum successful flights on the foam planes first before the heavier airframes can test high risk items - irrelevant for here but log does show 66 knots ground speed in the foam plane). 

I won’t have anyone today - Easter (which is too bad since it’s only 4 mph right now - granted it’s 4:30 am) and Monday - Wednesday says thunderstorms. I’ll look for a window and go if possible. We’ve got a foam and composite test plane queued up, just need a minute. If someone can go, me included I’ll get there.

**Comment by LorenzMeier** [comment]
@ryanjAA @dagar This is starting to become the last issue blocking the release.

**Comment by ryanjAA** [comment]
@LorenzMeier I don't think I'll be able to get anyone out to test this timely since most the testing capable team is heading to different locations shortly for various things occurring right now. If something changes quickly and we can deploy someone I'll write a note with the log attached. We still need to confirm whether this follows what is above from an absolute position or what you wrote in the slack fixedwing channel (mainly point 1):

1) Pitch offset is applied to everything now
2) The takeoff pitch is going to be exclusively used from the new parameter
3) You need to set pitch to zero in QGC for the plane to accept the mission


or are all now absolute from the horizon leveling with FW_PSP_OFF as essentially "cruise trim"?  In other words, takeoff pitch (from the parameter) is going to be absolute from the horizon leveled (not with the pitch offset added) and pitch needs to be set to 0 for the takeoff waypoint (in QGC mission planning) to accept the mission. This would make me assume that with a -5 PSP offset, a 3º min flare is 3º or rather than needing an 8º min flare (as it currently is) is needed to achieve a 3º min flare.

**Comment by dagar** [comment]
I've discussed this with @ryanjAA and we're going to bump this until after the v1.9.0 when we'll have more time to test and review.

**Comment by LorenzMeier** [comment]
Let's get this in! @dagar Could you rebase?

**Comment by ryanjAA** [comment]
I don’t think anyone has test flown this.

**Comment by Antiheavy** [comment]
@Kjkinney can we build this upstream PR into a branch and flight test this week?

**Comment by Kjkinney** [comment]
@Antiheavy done, should be good to test fly.

**Comment by Kjkinney** [comment]
Here is a flight log with this branch merged into the current 1.9.0 stable: 
https://review.px4.io/plot_app?log=7b9f80e7-b14b-40e6-b059-04f1b6d786c6

The branch it was tested on is slightly modified with some small custom changes on our end. 

For this flight I had FW_PSP_OFF set to 3, and FW_TKO_PITCH_MIN set to 25. The flight profile was a hand launched takeoff with a small box pattern at 200ft and a loiter to land. 

FYI @Antiheavy

**Comment by ryanjAA** [comment]
Question about this then. From the log, it shows that when you are in a landed state the Pitch SP is still +3º. Shouldn't that drop back to zero if this is only supposed to apply to pitch when at cruise and everything else is relative to (essentially) 0 deg, i.e. the initial setpoint to feed into TECS during cruise only? Reason being that if I open a different log with an offset (without the PR) and look at it, I see the offset pre and post takeoff, same as this log and then the flight tracks the pitch sp during the flight (as expected), so what are we seeing here that isnt the same as before?

Screen Shot 2019-05-30 at 1 17 42 PM

**Comment by Antiheavy** [comment]
@dagar should this newer PR be added to the Fixed Wing Project board in replacement of https://github.com/PX4/Firmware/pull/9243 ?

**Comment by Antiheavy** [comment]
> Question about this then. From the log, it shows that when you are in a landed state the Pitch SP is still +3º. Shouldn't that drop back to zero if this is only supposed to apply to pitch when at cruise and everything else is relative to (essentially) 0 deg, i.e. the initial setpoint to feed into TECS during cruise only? Reason being that if I open a different log with an offset (without the PR) and look at it, I see the offset pre and post takeoff, same as this log and then the flight tracks the pitch sp during the flight (as expected), so what are we seeing here that isnt the same as before?

@ryanjAA I'm not totally sure, but it is very reasonable that our aircraft (belly land foamy) was sitting at a different angle and spot on the ground after landing compared to before takeoff.  This improvement would be good to get in for v1.10 assuming it addresses all concerns.  Have you had chance to test independently?

**Comment by Antiheavy** [comment]
@LorenzMeier @dagar this would be a good candidate for 1.10 if we can bring it over the fence?  
Our main point of interest here is the new parameter FW_TKO_PITCH_MIN with which we've been flying for months now with great effect (our version is based on an old dead PR https://github.com/PX4/Firmware/pull/11373). 

FYI @Kjkinney

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 1]
```diff
@@ -1630,11 +1630,11 @@ float
 FixedwingPositionControl::get_tecs_pitch()
 {
 	if (_is_tecs_running) {
-		return _tecs.get_pitch_setpoint();
+		return _tecs.get_pitch_setpoint() + _parameters.pitchsp_offset_rad;
 	}
 
 	// return 0 to prevent stale tecs state when it's not running
-	return 0.0f;
+	return _parameters.pitchsp_offset_rad;
```
Comment (CarlOlsson) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: The comment above will need to be updated

**Comment by mrpollo** [comment]
This came up during the dev-call October 9th, 2019.

@dagar can you please rebase? looks like we need some test flight

**Comment by dagar** [comment]
@RomanBapst @sfuhrer perhaps we could talk this one through briefly on the next dev call? It shouldn't technically be that difficult, I just want to make sure we're all on the same page.

**Comment by RomanBapst** [comment]
@dagar @sfuhrer Ok, let's quickly discuss this so that we are on the same page and bring it in. I had a quick look and it makes sense to me. The way I see things:

All pitch limits should be absolute, when you tune TECS you are reading the absolute pitch angle, so it makes sense to have it this way.

**Comment by RomanBapst** [comment]
@dagar Do you have anything here that is not right? It all makes sense to me, I strongly agree to specify pitch limits globally always and also respect them always regardless of any offsets.
I guess this then just needs a rebase?

**Comment by dagar** [comment]
> @dagar Do you have anything here that is not right? It all makes sense to me, I strongly agree to specify pitch limits globally always and also respect them always regardless of any offsets.
> I guess this then just needs a rebase?

Yes, let's get this over the line.

**Comment by sfuhrer** [comment]
Closing in favor of https://github.com/PX4/PX4-Autopilot/pull/16792

**Pull request closed without merging**